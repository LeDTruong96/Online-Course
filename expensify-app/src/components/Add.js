import React from 'react';

function AddExpensePage() {
	return (
		<div>
			<p>This is my add page</p>
		</div>
	);
}

export default AddExpensePage;
