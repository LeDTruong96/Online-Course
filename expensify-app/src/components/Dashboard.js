import React from 'react';

function DashboardPage() {
	return (
		<div>
			<p>This is my home page</p>
		</div>
	);
}

export default DashboardPage;
