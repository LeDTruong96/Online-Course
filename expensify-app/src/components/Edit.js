import React from 'react';

function EditExpensePage(props) {
	console.log(props);
	return (
		<div>
			<p>This is my Edit page</p>
		</div>
	);
}

export default EditExpensePage;
