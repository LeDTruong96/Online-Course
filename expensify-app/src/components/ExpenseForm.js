import React, { Component } from 'react';

export default class ExpenseForm extends Component {
	state = {
		description: '',
		amount: '',
		note: ''
	};

	onDescriptionChange = (e) => {
		this.setState(() => {
			return { description: e.target.value };
		});
	};

	onAmountChange = (e) => {
		console.log(!!e.target.value.match(/[\d]*/)[0]);
		if (!!e.target.value.match(/[\d]*/)[0]) {
			this.setState(() => {
				return { amount: e.target.value };
			});
		}
	};

	onNoteChange = (e) => {
		this.setState(() => {
			return { note: e.target.value };
		});
	};

	render() {
		return (
			<div>
				<form action="">
					<input
						type="text"
						placeholder="description"
						autoFocus={true}
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input
						type="text"
						placeholder="Amount"
						value={this.state.amount}
						onChange={this.onAmountChange}
					/>
					<textarea
						placeholder="Add your notes in here (option)"
						value={this.state.note}
						onChange={this.onNoteChange}
					/>
					<button>Add Expense</button>
				</form>
			</div>
		);
	}
}
