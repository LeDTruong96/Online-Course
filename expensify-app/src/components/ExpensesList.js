import React from 'react';
import { connect } from 'react-redux';
import ExpensesListItem from './ExpensesListItem';
import getVisiableExpenese from '../selectors/expenses';

const ExpensesList = (props) => {
	return (
		<div>
			<h1>EXPENSES LIST</h1>
			{props.expenses.map((expense) => {
				return <ExpensesListItem key={expense.id} {...expense} dispatch={props.dispatch} />;
			})}
		</div>
	);
};

const getPropsFromStore = (props) => {
	return {
		expenses: getVisiableExpenese(props)
	};
};

export default connect(getPropsFromStore)(ExpensesList);
