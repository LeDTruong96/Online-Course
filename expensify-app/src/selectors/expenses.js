// Get visible expenses
export default ({ expenses, filters }) => {
	return expenses
		.filter((expense) => {
			const matchStartDate =
				typeof filters.startDate !== 'number' || expense.createAt >= filters.startDate;
			const matchEndDate =
				typeof filters.endDate !== 'number' || expense.createAt <= filters.endDate;
			const matchText = expense.description
				.toLowerCase()
				.includes(filters.text.toLowerCase());

			return matchStartDate && matchEndDate && matchText;
		})
		.sort((fExpense, sExpense) => {
			if (filters.sortBy === 'date') {
				return fExpense.createAt < sExpense.createAt ? 1 : -1;
			} else if (filters.sortBy === 'amount') {
				return fExpense.amount < sExpense.amount ? 1 : -1;
			}
		});
};
