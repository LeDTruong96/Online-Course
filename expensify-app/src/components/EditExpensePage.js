import React, { Component } from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends Component {
	onSubmit = (id, expense) => {
		this.props.editExpense(id, expense);
		this.props.history.push('/');
	};
	onClick = () => {
		this.props.removeExpense(this.props.expense.id);
		this.props.history.push('/');
	};

	render() {
		return (
			<div>
				<ExpenseForm
					type={'Save Expense'}
					Expense={this.props.expense}
					onSubmit={this.onSubmit}
				/>
				<button onClick={this.onClick}>Remove Expense </button>
			</div>
		);
	}
}

const getExpenseFromStore = (states, props) => {
	return {
		expense: states.expenses.find((expense) => {
			return expense.id === props.match.params.id;
		})
	};
};

const mapDispatchToProps = (dispatch, props) => {
	return {
		editExpense: () => {
			dispatch(editExpense(props.expense.id, expense));
		},
		removeExpense: () => {
			dispatch(removeExpense(props.expense.id));
		}
	};
};
export default connect(getExpenseFromStore, mapDispatchToProps)(EditExpensePage);
