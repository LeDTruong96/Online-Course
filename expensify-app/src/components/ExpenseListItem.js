import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
	return (
		<div>
			<Link to={`/edit/${id}`}>
				<h3>{description}</h3>
			</Link>
			<p>
				{amount} - {moment(createdAt).format('MMM Do, YYYY')}
			</p>
		</div>
	);
};

export default ExpenseListItem;
