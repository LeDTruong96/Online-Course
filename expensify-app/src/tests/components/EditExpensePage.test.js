import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense, removeExpense, history, match, wrapper;

beforeEach(() => {
	editExpense = jest.fn();
	removeExpense = jest.fn();
	history = { push: jest.fn() };
	wrapper = shallow(
		<EditExpensePage
			editExpense={editExpense}
			history={history}
			match={match}
			expense={expenses[0]}
			removeExpense={removeExpense}
		/>
	);
});

test('EditExpensePage match the snapshot', () => {
	expect(wrapper).toMatchSnapshot();
});

test('EditExpensePage edit expense', () => {
	wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0].id, expenses[0]);
	expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
	expect(history.push).toHaveBeenLastCalledWith('/');
});

test('EditExpensePage edit expense', () => {
	wrapper.find('button').simulate('click');
	expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
});
