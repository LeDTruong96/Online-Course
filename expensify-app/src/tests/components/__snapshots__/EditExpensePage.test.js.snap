// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage match the snapshot 1`] = `
<div>
  <ExpenseForm
    Expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
    type="Save Expense"
  />
  <button
    onClick={[Function]}
  >
    Remove Expense 
  </button>
</div>
`;
